include/boost/math/special_functions/prime_sieve.hpp

```diff
@@ -0,0 +1,377 @@
+    {
+        boost::math::detail::segmented_sieve(static_cast<Z>(2), upper_bound, primes);
+    }
+
+    else
+    {
+        unsigned processor_count {std::thread::hardware_concurrency()};
+
+        // May return 0 when unable to detect
+        if(processor_count == 0)
+        {
+            processor_count = 2;
+        }
+
+        std::vector<Z> small_primes {};
+        small_primes.reserve(1000);
+
+        // Threshold for when 2 thread performance begins to be non-linear, or when the system can only support two threads
+        if(upper_bound <= 16777216 || processor_count == 2)
+        {
+            // Split into two vectors and merge after joined to avoid data races
+            std::thread t1([&small_primes] {
+                boost::math::detail::linear_sieve(static_cast<Z>(8192), small_primes);
+            });
+            std::thread t2([upper_bound, &primes] {
+                boost::math::detail::segmented_sieve(static_cast<Z>(8192), upper_bound, primes);
+            });
+
+            t1.join();
+            t2.join();
+            primes.insert(primes.begin(), small_primes.begin(), small_primes.end());
+        }
+
+        //If sufficiently large upper bound spawn as many threads as the system has processors for
+        else
+        {
+            //Pre-generate all of the primes so that each thread does not have to
+            Z limit{static_cast<Z>(std::floor(std::sqrt(static_cast<double>(upper_bound)))) + 1};
+            std::vector<Z> pre_generated_primes {};
+            pre_generated_primes.reserve(limit / std::log(limit));
+
+            if(limit < 8192)
+            {
+                boost::math::detail::linear_sieve(limit, pre_generated_primes);
+            }
+
+            else
+            {
+                std::thread t1([&small_primes] {
+                    boost::math::detail::linear_sieve(static_cast<Z>(8192), small_primes);
+                });
+                std::thread t2([limit, &pre_generated_primes] {
+                    boost::math::detail::segmented_sieve(static_cast<Z>(8192), limit, pre_generated_primes);
+                });
+
+                t1.join();
+                t2.join();
+                pre_generated_primes.insert(pre_generated_primes.begin(), small_primes.begin(), small_primes.end());
+            }
+
+            std::vector<std::thread> thread_manager {};
+            std::vector<std::vector<Z>> prime_vectors(processor_count - 1);
+            const Z range_per_thread = upper_bound / (processor_count - 1);
+            Z current_lower_bound {8192};
+            Z current_upper_bound {current_lower_bound + range_per_thread};
+            Z primes_in_range {static_cast<Z>(current_upper_bound / std::log(static_cast<double>(current_upper_bound)) -
+                               current_lower_bound / std::log(static_cast<double>(current_lower_bound)))};
+
+            for(size_t i{1}; i < processor_count - 1; ++i)
+            {
+                std::vector<Z> temp {};
```

Comment:
I don't think this is doing what you think it is, and in any case is the cause of the segfaults I'm seeing, as uninitialized vectors are being passed to semented_sieve.

First you need 

```
std::vector<std::vector<Z>> prime_vectors(processor_count);
```

above (count them!), then in the loop you should remove `temp` and just call:

```
prime_vectors[i - 1].reserve(primes_in_range);
```

No emplace_back required.

Then update the index used in the lambda to i - 1 as well.  Or else just start from zero!!

Then the final N'th thread created outside the loop needs the same fix.